deepinv/training_utils.py

```diff
@@ -73,130 +70,196 @@ def train(
     :param bool unsupervised: Train an unsupervised network, i.e., uses only measurement vectors y for training.
     :param bool plot_images: Plots reconstructions every ``ckp_interval`` epochs.
     :param bool wandb_vis: Use Weights & Biases visualization, see https://wandb.ai/ for more details.
+    :param dict wandb_setup: Dictionary with the setup for wandb, see https://docs.wandb.ai/quickstart for more details.
+    :param bool online_measurements: Generate the measurements in an online manner at each iteration by calling
+        ``physics(x)``. This results in a wider range of measurements if the physics' parameters, such as
+         parameters of the forward operator or noise realizations, can change between each sample; these are updated
+         with the ``physics.reset()`` method. If ``online_measurements=False``, the measurements are generated
+        offline before training and saved in memory.
     :returns: Trained model.
     """
     save_path = Path(save_path)
 
+    # wandb initialiation
     if wandb_vis:
         if wandb.run is None:
-            wandb.init()
+            wandb.init(**wandb_setup)
 
+    # set the different metrics
+    meters = []
+    total_loss = AverageMeter("loss", ":.2e")
+    meters.append(total_loss)
     if not isinstance(losses, list) or isinstance(losses, tuple):
         losses = [losses]
-
-    loss_meter = AverageMeter("loss", ":.2e")
-    meters = [loss_meter]
-    eval_psnr_net = []
-
     losses_verbose = [AverageMeter("Loss_" + l.name, ":.2e") for l in losses]
-    train_psnr_net = AverageMeter("Train_psnr_model", ":.2f")
-    if eval_dataloader:
-        eval_psnr_net = AverageMeter("Eval_psnr_model", ":.2f")
-
     for loss in losses_verbose:
         meters.append(loss)
-
-    meters.append(train_psnr_net)
+    train_psnr = AverageMeter("Train_psnr_model", ":.2f")
+    meters.append(train_psnr)
     if eval_dataloader:
-        meters.append(eval_psnr_net)
-
-    progress = ProgressMeter(epochs, meters)
+        eval_psnr = AverageMeter("Eval_psnr_model", ":.2f")
+        meters.append(eval_psnr)
 
     save_path = f"{save_path}/{get_timestamp()}"
 
+    # count the overall training parameters
     params = sum(p.numel() for p in model.parameters() if p.requires_grad)
     print(f"The model has {params} trainable parameters")
 
+    # make physics and data_loaders of list type
     if type(physics) is not list:
         physics = [physics]
-
-    if type(losses) is not list:
-        losses = [losses]
-
     if type(train_dataloader) is not list:
         train_dataloader = [train_dataloader]
-
     if eval_dataloader and type(eval_dataloader) is not list:
         eval_dataloader = [eval_dataloader]
 
     G = len(train_dataloader)
 
     loss_history = []
 
+    log_dict = {}
+
     for epoch in range(epochs):
+        ### Evaluation
+
+        # perform evaluation every eval_interval epoch
+        perform_eval = (
```

Comment:
the evaluation shouldn't happen after the training step inside the epoch?

Reply:
I think it is usual to do an eval step before all so that it tests the initialized network and also be sure that there is no error in the eval loop without wiating for a whole trianing epoch

Reply:
good point, I agree